File: 历史版本/C_Method2026.5.20/F_series_gen.py
import numpy as np

def F_series_gen(a_fun,m,T,nDim,cut_small=True):
    '''
    profile:光栅轮廓函数,m:FFT级次,T:周期长度,nDim:模态数
    该函数返回-N/2~N/2级谐波
    '''
    N=2**m
    tol=1e-9
    x=np.linspace(0,T,N+1)
    f=a_fun(x)
    fhat=f
    fhat[0]=(f[0]+f[-1])/2
    fhat=fhat[:-1]
    fourier_coeffi=np.fft.fftshift(np.fft.fft(fhat)/fhat.shape[0])
    #减去系数中的小量
    if cut_small==True:
        ind_small_real=(np.abs(np.real(fourier_coeffi))<tol)
        fourier_coeffi[ind_small_real]=1j*np.imag(fourier_coeffi[ind_small_real])
        ind_small_imag=(np.abs(np.imag(fourier_coeffi))<tol)
        fourier_coeffi[ind_small_imag]=np.real(fourier_coeffi[ind_small_imag])
    q0=len(fourier_coeffi)//2
    return fourier_coeffi[q0-nDim:q0+nDim+1]

File: 现有库/Chandezon-Method-master/F_series_gen.m
%% Generates Fouries series for given periodic function

function[f_vec] =  F_series_gen(a_fun,m,T,nDim) 
%input parameters - function profile, order of fft, period
%length, number of terms in the series
N=2^m;  
tol=10e-10; % treshold for cutting small array elements
cut_small=1; % cut small array elements
y_vec=linspace(0,T,N+1); % prepare nodes
f = a_fun(y_vec);   % function values in the nodes
fhat =f;   
fhat(1)=(f(1)+f(N+1))/2;            % average of the first and last element - for faster convergence
fhat(N+1)=[];

F=fft(fhat,N)/N;    % do FFT
% cut small array elements
if (cut_small==1)
    ind_small_real=(abs(real(F))<tol);
    F(ind_small_real)=1i.*imag(F(ind_small_real));
    ind_small_imag=(abs(imag(F))<tol);
    F(ind_small_imag)=real(F(ind_small_imag));
    f_vec=F(1:nDim)';
%    f_vec=f_vec(1:2:end);
end